### PR TITLE
Bulk index findings and sequentially invoke auto-correlations

### PR DESCRIPTION
**Pull request by goyamegh**
*Issue #, if available:*
#1333 

*Description of changes:*
All the findings once generated, are indexed one-by-one today. This PR attempts to optimize this by making a bulk index request for all the generated findings. Once completed, each finding will then be published to kick-start auto-correlations subsequently.

*CheckList:*
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/alerting/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: alerting/src/main/kotlin/org/opensearch/alerting/settings/AlertingSettings.kt [hunk 1]
```diff
@@ -17,6 +17,7 @@ class AlertingSettings {
 
     companion object {
         const val DEFAULT_MAX_ACTIONABLE_ALERT_COUNT = 50L
+        const val DEFAULT_FINDINGS_INDEXING_BATCH_SIZE = 10000
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/settings/AlertingSettings.kt:20]: better to be 1000.

File: alerting/src/main/kotlin/org/opensearch/alerting/settings/AlertingSettings.kt [hunk 2]
```diff
@@ -153,5 +154,12 @@ class AlertingSettings {
             -1L,
             Setting.Property.NodeScope, Setting.Property.Dynamic
         )
+
+        val FINDINGS_INDEXING_BATCH_SIZE = Setting.intSetting(
+            "plugins.alerting.alert_findings_indexing_batch_size",
+            DEFAULT_FINDINGS_INDEXING_BATCH_SIZE,
+            0,
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/settings/AlertingSettings.kt:161]: min cant be 0

**Review by eirsep — CHANGES_REQUESTED** [review]

File: alerting/src/main/kotlin/org/opensearch/alerting/AlertingPlugin.kt [hunk 3]
```diff
@@ -346,7 +346,8 @@ internal class AlertingPlugin : PainlessExtension, ActionPlugin, ScriptPlugin, R
             AlertingSettings.FINDING_HISTORY_MAX_DOCS,
             AlertingSettings.FINDING_HISTORY_INDEX_MAX_AGE,
             AlertingSettings.FINDING_HISTORY_ROLLOVER_PERIOD,
-            AlertingSettings.FINDING_HISTORY_RETENTION_PERIOD
+            AlertingSettings.FINDING_HISTORY_RETENTION_PERIOD,
+            AlertingSettings.FINDINGS_INDEXING_BATCH_SIZE
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/AlertingPlugin.kt:350]: nit:BULK

File: alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt [hunk 4]
```diff
@@ -461,51 +460,102 @@ object DocumentLevelMonitorRunner : MonitorRunner() {
...
-            monitorName = monitor.name,
-            index = docIndex[1],
-            docLevelQueries = docLevelQueries,
-            timestamp = Instant.now(),
-            executionId = workflowExecutionId
-        )
+        val findingDocPairs = mutableListOf<Pair<String, String>>()
+        val findings = mutableListOf<Finding>()
+        val indexRequests = mutableListOf<IndexRequest>()
+        monitorCtx.findingsIndexBatchSize = FINDINGS_INDEXING_BATCH_SIZE.get(monitorCtx.settings)
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt:480]: what is this - `monitorCtx.findingsIndexBatchSize`??

Why are we mutating value of monitorCtx object?? that object should only be read from

**Comment by eirsep** [comment]
can we test the new setting?

File: alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt [hunk 5]
```diff
@@ -461,51 +460,102 @@ object DocumentLevelMonitorRunner : MonitorRunner() {
...
+        monitorCtx.clusterService!!.clusterSettings.addSettingsUpdateConsumer(FINDINGS_INDEXING_BATCH_SIZE) {
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt:481]: we aren't supposed to add settings consumer in the middle of monitor execution. plz do at node startup

File: alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt [hunk 6]
```diff
@@ -461,51 +460,102 @@ object DocumentLevelMonitorRunner : MonitorRunner() {
...
+    }
+
+    private suspend fun bulkIndexFindings(
+        monitor: Monitor,
+        monitorCtx: MonitorRunnerExecutionContext,
+        indexRequests: List<IndexRequest>
+    ) {
+        if (indexRequests.isNotEmpty()) {
+            val bulkResponse: BulkResponse = monitorCtx.client!!.suspendUntil {
+                bulk(BulkRequest().add(indexRequests).setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE), it)
```
Comment (eirsep) [inline alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt:547]: dont refresh for every batch. refresh findings index only once after all batches are bulk-ingested

File: alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt (hunk 5 above)
Comment (goyamegh) [reply alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt:481]: Moved it to registerConsumers() in MonitorRunnerService.

File: alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt (hunk 4 above)
Comment (goyamegh) [reply alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt:480]: The intent is to update this value in alerting, when it is updated via _cluster/settings.

File: alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt [hunk 7]
```diff
@@ -374,35 +373,34 @@ object DocumentLevelMonitorRunner : MonitorRunner() {
...
+            // Only pick those entries whose docs have triggers associated with them
+            if (triggerResult.triggeredDocs.contains(it.second)) {
+                triggerFindingDocPairs.add(Pair(it.first, it.second))
             }
         }
 
         val actionCtx = triggerCtx.copy(
             triggeredDocs = triggerResult.triggeredDocs,
-            relatedFindings = findings,
+            // confirm if this is right or only trigger-able findings should be present in this list
```
[the remark this answers is not in the document]
Comment (goyamegh) [reply alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt:398]: This change is preserving the existing behavior. Left by mistake, removed the comment.

File: alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt (hunk 6 above)
Comment (goyamegh) [reply alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt]: Done.

**Review by eirsep — COMMENTED** [review]
plz add a test where you make bulk finding size =2 and simulate creation of 10 findings

**Review by eirsep — APPROVED** [review]

**Review by eirsep — APPROVED** [review]

**Comment by eirsep** [comment]
tests are failing?

**Review by praveensameneni — APPROVED** [review]

**Review by eirsep — APPROVED** [review]

**Pull request merged**